src/Raven.Server/Documents/Replication/DocumentReplicationLoader.cs

```diff
@@ -1,118 +1,48 @@
-                }
-
-                Debug.Assert(_replicationDocument.Destinations != null);
-                OnConfigurationChanged(_replicationDocument.Destinations);
-            }
-        }
-
-        //TODO: add here error handling for the following cases
-        //1) what if unexpected exception happens in outgoing replication dispose?
-        //2) what if sending a replication batch is happening during a call to dispose?
-        protected void OnConfigurationChanged(List<ReplicationDestination> destinations)
-        {
-            lock (_outgoingReplications)
-            {
-                foreach (var replication in _outgoingReplications)
-                    replication.Dispose();
-                _outgoingReplications.Clear();
-
-                var initializationTasks = new List<Task>();
-                foreach (var dest in destinations)
-                {
-                    var outgoingDocumentReplication = new OutgoingDocumentReplication(_database, dest);
-                    initializationTasks.Add(outgoingDocumentReplication.InitializeAsync());
-                    _outgoingReplications.Add(outgoingDocumentReplication);
-                }
-
-                Task.WhenAll(initializationTasks).Wait(_database.DatabaseShutdown);
-            }
-        }
-
-        public void Dispose()
-        {
-            lock (_outgoingReplications)
-            {
-                foreach (var replication in _outgoingReplications)
-                    replication.Dispose();
-            }
-            _database.Notifications.OnSystemDocumentChange -= HandleSystemDocumentChange;
-        }
-
-        private void HandleSystemDocumentChange(DocumentChangeNotification notification)
-        {
-            if (ShouldReloadConfiguration(notification.Key))
-            {
-                LoadConfigurations();
-
-                if (_log.IsDebugEnabled)
-                    _log.Debug($"Replication configuration was changed: {notification.Key}");
-            }
-        }
-    }
+	public class DocumentReplicationLoader : IDisposable
+	{
+		private readonly DocumentDatabase _database;
+
+		private readonly ConcurrentDictionary<ReplicationDestination,OutgoingReplicationHandler> _outgoing = new ConcurrentDictionary<ReplicationDestination, OutgoingReplicationHandler>();
+		private readonly ConcurrentDictionary<TcpConnectionHeaderMessage, IncomingReplicationHandler> _incoming = new ConcurrentDictionary<TcpConnectionHeaderMessage, IncomingReplicationHandler>();
+
+		public DocumentReplicationLoader(DocumentDatabase database)
+		{
+			_database = database;
+		}
+
+		public void AcceptIncomingConnection(TcpConnectionHeaderMessage incomingMessageHeader, NetworkStream tcpStream)
+		{
+			_incoming.AddOrUpdate(incomingMessageHeader,
+				key =>
+				{
+					var newIncoming = new IncomingReplicationHandler(key, tcpStream, _database);
+					newIncoming.Start();
+					return newIncoming;
+				},
+				(key, existing) =>
+				{
+					existing.Dispose();
```

Comment:
Logging is required here.
And need to also report it to the other side.

However, this means that you can have just a single replication connection to a database.
You are comparing the _incoming connection details_, not the outgoing connection